### PR TITLE
66, 77 and 80 - Added support for secondary index, materialized view and UDF/UDA metadata

### PR DESCRIPTION
**Pull request by mpenick**

### CODE REVIEW

File: ext/src/Cassandra/DefaultKeyspace.c [hunk 1]
```diff
@@ -195,10 +208,286 @@ PHP_METHOD(DefaultKeyspace, userTypes)
...
+    } else {
+      view = PHP_CASSANDRA_GET_MATERIALIZED_VIEW(PHP5TO7_ZVAL_MAYBE_P(zview));
+
+      if (PHP5TO7_Z_TYPE_MAYBE_P(view->name) == IS_STRING) {
+        PHP5TO7_ADD_ASSOC_ZVAL_EX(return_value,
+                                  PHP5TO7_Z_STRVAL_MAYBE_P(view->name),
+                                  PHP5TO7_Z_STRLEN_MAYBE_P(view->name) + 1,
+                                  PHP5TO7_ZVAL_MAYBE_P(zview));
+      } else {
+        add_next_materialized_view_zval(return_value, PHP5TO7_ZVAL_MAYBE_P(zview));
```
Comment (mikefero) [inline ext/src/Cassandra/DefaultKeyspace.c]: `add_next_materialized_view_zval` is undefined

File: ext/src/Cassandra/DefaultKeyspace.c [hunk 2]
```diff
@@ -195,10 +208,286 @@ PHP_METHOD(DefaultKeyspace, userTypes)
...
+    if (!PHP5TO7_ZVAL_IS_UNDEF(zfunction)) {
+      cassandra_function *function = PHP_CASSANDRA_GET_FUNCTION(PHP5TO7_ZVAL_MAYBE_P(zfunction));
+
+      if (PHP5TO7_Z_TYPE_MAYBE_P(function->signature) == IS_STRING) {
+        PHP5TO7_ADD_ASSOC_ZVAL_EX(return_value,
+                                  PHP5TO7_Z_STRVAL_MAYBE_P(function->signature),
+                                  PHP5TO7_Z_STRLEN_MAYBE_P(function->signature) + 1,
+                                  PHP5TO7_ZVAL_MAYBE_P(zfunction));
+      } else {
+        add_next_function_zval(return_value, PHP5TO7_ZVAL_MAYBE_P(zfunction));
```
Comment (mikefero) [inline ext/src/Cassandra/DefaultKeyspace.c]: `add_next_function_zval` is undefined

File: ext/src/Cassandra/DefaultKeyspace.c [hunk 3]
```diff
@@ -195,10 +208,286 @@ PHP_METHOD(DefaultKeyspace, userTypes)
...
+    if (!PHP5TO7_ZVAL_IS_UNDEF(zaggregate)) {
+      cassandra_aggregate *aggregate = PHP_CASSANDRA_GET_AGGREGATE(PHP5TO7_ZVAL_MAYBE_P(zaggregate));
+
+      if (PHP5TO7_Z_TYPE_MAYBE_P(aggregate->signature) == IS_STRING) {
+        PHP5TO7_ADD_ASSOC_ZVAL_EX(return_value,
+                                  PHP5TO7_Z_STRVAL_MAYBE_P(aggregate->signature),
+                                  PHP5TO7_Z_STRLEN_MAYBE_P(aggregate->signature) + 1,
+                                  PHP5TO7_ZVAL_MAYBE_P(zaggregate));
+      } else {
+        add_next_aggregate_zval(return_value, PHP5TO7_ZVAL_MAYBE_P(zaggregate));
```
Comment (mikefero) [inline ext/src/Cassandra/DefaultKeyspace.c]: `add_next_aggregate_zval` is undefined

File: ext/src/Cassandra/DefaultTable.c [hunk 4]
```diff
@@ -331,60 +398,218 @@ PHP_METHOD(DefaultTable, columns)
...
+    if (!PHP5TO7_ZVAL_IS_UNDEF(zview)) {
+      view = PHP_CASSANDRA_GET_MATERIALIZED_VIEW(PHP5TO7_ZVAL_MAYBE_P(zview));
+
+      if (PHP5TO7_Z_TYPE_MAYBE_P(view->name) == IS_STRING) {
+        PHP5TO7_ADD_ASSOC_ZVAL_EX(return_value,
+                                  PHP5TO7_Z_STRVAL_MAYBE_P(view->name),
+                                  PHP5TO7_Z_STRLEN_MAYBE_P(view->name) + 1,
+                                  PHP5TO7_ZVAL_MAYBE_P(zview));
+      } else {
+        add_next_materialized_view_zval(return_value, PHP5TO7_ZVAL_MAYBE_P(zview));
```
Comment (mikefero) [inline ext/src/Cassandra/DefaultTable.c]: `add_next_materialized_view_zval` is undefined

File: ext/src/Cassandra/DefaultMaterializedView.c [hunk 5]
```diff
@@ -0,0 +1,632 @@
...
+  PHP_ME(DefaultMaterializedView, compactionStrategyClassName, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultMaterializedView, compactionStrategyOptions, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultMaterializedView, compressionParameters, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultMaterializedView, populateIOCacheOnFlush, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultMaterializedView, replicateOnWrite, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultMaterializedView, maxIndexInterval, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultMaterializedView, minIndexInterval, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultMaterializedView, column, arginfo_name, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultMaterializedView, columns, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultMaterializedView, primaryKey, arginfo_none, ZEND_ACC_PUBLIC)
```
Comment (mikefero) [inline ext/src/Cassandra/DefaultMaterializedView.c:570]: This diverts from the naming in the cpp-driver a bit; however `primaryKey` is clearer. Maybe we could add an alias for `primaryKey` called `partitionKey`.

File: ext/src/Cassandra/Aggregate.c [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "php_cassandra.h"
+
+zend_class_entry *cassandra_aggregate_ce = NULL;
+
+ZEND_BEGIN_ARG_INFO_EX(arginfo_none, 0, ZEND_RETURN_VALUE, 0)
+ZEND_END_ARG_INFO()
+
+static zend_function_entry cassandra_aggregate_methods[] = {
+  PHP_ABSTRACT_ME(Aggregate, simpleName, arginfo_none)
```
Comment (mikefero) [inline ext/src/Cassandra/Aggregate.c:26]: `name` would keep consistency with the cpp-driver

File: ext/src/Cassandra/Aggregate.c [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  PHP_ABSTRACT_ME(Aggregate, argumentTypes, arginfo_none)
+  PHP_ABSTRACT_ME(Aggregate, finalFunction, arginfo_none)
+  PHP_ABSTRACT_ME(Aggregate, stateFunction, arginfo_none)
+  PHP_ABSTRACT_ME(Aggregate, initialCondition, arginfo_none)
+  PHP_ABSTRACT_ME(Aggregate, returnType, arginfo_none)
+  PHP_ABSTRACT_ME(Aggregate, stateType, arginfo_none)
+  PHP_ABSTRACT_ME(Aggregate, signature, arginfo_none)
```
Comment (mikefero) [inline ext/src/Cassandra/Aggregate.c:33]: Same here ... should we create an alias called `fullName` for consistency?

File: ext/src/Cassandra/DefaultAggregate.c [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+
+  self = PHP_CASSANDRA_GET_AGGREGATE(getThis());
+  RETURN_ZVAL(PHP5TO7_ZVAL_MAYBE_P(self->signature), 1, 0);
+}
+
+ZEND_BEGIN_ARG_INFO_EX(arginfo_none, 0, ZEND_RETURN_VALUE, 0)
+ZEND_END_ARG_INFO()
+
+static zend_function_entry cassandra_default_aggregate_methods[] = {
+  PHP_ME(DefaultAggregate, simpleName, arginfo_none, ZEND_ACC_PUBLIC)
```
Comment (mikefero) [inline ext/src/Cassandra/DefaultAggregate.c:201]: Same as with aggregate

File: ext/src/Cassandra/DefaultAggregate.c [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+  PHP_ME(DefaultAggregate, argumentTypes, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultAggregate, stateFunction, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultAggregate, finalFunction, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultAggregate, initialCondition, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultAggregate, stateType, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultAggregate, returnType, arginfo_none, ZEND_ACC_PUBLIC)
+  PHP_ME(DefaultAggregate, signature, arginfo_none, ZEND_ACC_PUBLIC)
```
Comment (mikefero) [inline ext/src/Cassandra/DefaultAggregate.c:208]: as well as here

**Comment by mikefero** [comment]
All new features are missing documentation stubs but other than this and the comments/suggestions everything looks good and is tested.

**Comment by mpenick** [comment]
Rebased with `master` and merged at 100043602ada75802ee893c32c8bcf39c999b950

**Pull request closed without merging**